### ISSUE TITLE
Problems cmorizing IFS output

### ISSUE THREAD

**Issue by LaKenLin** [EC-Earth/ece2cmor3#703]
Hi!

I am working on cmorizing IFS output with ece2cmor v1.7.0. Only for the first year, I get problems with a handful of variables.
When I use mrsos/3hr and tsl/6hrPlevPt in the variable list, I get error messages looking like the following for both of them:

> ^[[2;31;47mC Traceback:
> ! In function: cmor_check_monotonic
> ! called from: cmor_treat_axis_values
> ! called from: cmor_axis
> ! ^[[0m
> 
> ^[[1;31;47m!!!!!!!!!!!!!!!!!!!!!!!!!
> !
> ! Error: axis time1 (table: 3hr) has non monotonic values:
> ! 60265.000000, 60265.000000 and  60265.250000
> !
> !!!!!!!!!!!!!!!!!!!!!!!!!^[[0m

This is usually followed by the run continuing for a bit and then just stopping but not crashing without any apparent error message. 
I should mention that I tried to link for example ICMGG????+201501 as the previous month file for a run starting in 2015, rather than ICMGG????+000000 and this seemed to 'solve' the problem. Also, running the same year with ece2cmor v1.5 comepleted the year without this problem.

I am also having problems with cdo not being able to read some of the GRIB-files produced, throwing a "Unsupported file structure" error. This problem again only occurs for 2015 and affects the variables tslsi/3hr, tslsi/day, mrfso/Lmon and tsl/Lmon.
grib_ls shows, that there is one additional record in the file for 2015, compared to 2042:
```
grib_ls -x -P paramId -w typeOfLevel=surface,dataDate=20150101 2015---235.128.1_172.128.1.3 
2015---235.128.1_172.128.1.3
paramId      edition      centre       typeOfLevel  level        dataDate     stepRange    dataType     shortName    
172          1            ecmf         surface      0            20150101     0            fc           lsm         
172          1            ecmf         surface      0            20150101     0            fc           lsm         
235          1            ecmf         surface      0            20150101     0            fc           skt         
235          1            ecmf         surface      0            20150101     0            fc           skt         
235          1            ecmf         surface      0            20150101     0            fc           skt         
235          1            ecmf         surface      0            20150101     0            fc           skt         
235          1            ecmf         surface      0            20150101     0            fc           skt         
235          1            ecmf         surface      0            20150101     0            fc           skt         
235          1            ecmf         surface      0            20150101     0            fc           skt         
235          1            ecmf         surface      0            20150101     0            fc           skt 
```
```
grib_ls -x -P paramId -w typeOfLevel=surface,dataDate=20420101 2042---235.128.1_172.128.1.3 
2042---235.128.1_172.128.1.3
paramId      edition      centre       typeOfLevel  level        dataDate     stepRange    dataType     shortName    
172          1            ecmf         surface      0            20420101     0            fc           lsm         
235          1            ecmf         surface      0            20420101     0            fc           skt         
235          1            ecmf         surface      0            20420101     0            fc           skt         
235          1            ecmf         surface      0            20420101     0            fc           skt         
235          1            ecmf         surface      0            20420101     0            fc           skt         
235          1            ecmf         surface      0            20420101     0            fc           skt         
235          1            ecmf         surface      0            20420101     0            fc           skt         
235          1            ecmf         surface      0            20420101     0            fc           skt         
235          1            ecmf         surface      0            20420101     0            fc           skt
```
Could this be the issue?

Any ideas and help appreciated!

**Comment by treerink** [comment]
The fact that you can run with `ece2cmor3 v1.5` and that you have trouble with `v1.7` related with the ` ICMGG????+000000`, might have to do with the fact that between these versions the way grib code 172 is used changed. Could it be that in the file which you linked or used as ` ICMGG????+000000` 172 is missing?

**Comment by LaKenLin** [comment]
Code 172 is there:
```
grib_ls -x -P paramId ICMGGds11+000000 | grep 172
172          1            ecmf         surface      0            20150101     0            fc           lsm
```

**Comment by treerink** [comment]
Okay if 172 is in the ICMGG????+000000 then the `v1.5` - `v1.7` difference is more puzzling.

Just another basic check, if applicable in your case, you take care of this:
[Recommended strategies](https://github.com/EC-Earth/ece2cmor3/wiki/Recommended-strategies#recommended-strategies)

**Comment by goord** [comment]
Hi I think the duplicate GRIB message of the lsm in the filtered grib files is causing your problems. Something changed in the filtering of the lsm

**Comment by LaKenLin** [comment]
@teerink it looks like we are following the suggested procedure for restart experiments. x000000 files are linked as x184912 for example.

@goord Yes, it looks like that. At least, that is the only difference I can see between the first year of run file, compared to later years. Any suggestions on how to go about fixing this manually? Since the grib file is generated during the run and further processed after the failing cdo commands are applied, I cannot really imagine that fixing it manually can be done.

**Comment by goord** [comment]
Hi @LaKenLin, it cannot and shouldn't be fixed manually. I will have a look at it this afternoon. Are the grib files large? Could put the ICM{SH,GG}ds11+000000, ICM{SH,GG}ds11+201412, ICM{SH,GG}ds11+201501 on an ftp site or dropbox?

**Comment by LaKenLin** [comment]
I see. Thank you!
They are quite large, compressed together they have 9G. But you can find them here: https://www.dropbox.com/s/xhq8pq6swkky3uj/ds11_GG_SH.tar.gz?dl=0

**Comment by goord** [comment]
Hi @LaKenLin I'm having difficulties reproducing your problem, could you post your varlist.json as well?

**Comment by LaKenLin** [comment]
Hi @goord 
In this [directory](https://www.dropbox.com/sh/vag36xeb8wdvqkj/AACpkPaPRQ0KojNlpVRiJ7pua?dl=0) I placed a varlist with the variables causing the problem as well as an out-log, a run script and the metadata file.

**Comment by LaKenLin** [comment]
Hi @goord 
Could you have a quick look at the runscript and log I posted and let me know if you can see anything suspicious? Would be much appreciated!

**Comment by goord** [comment]
Hi @LaKenLin I just downloaded your run script and config, will have a look at it later today. I'll get back to you tomorrow

**Comment by LaKenLin** [comment]
Thank you!

My theory is that problems are caused by the fact that the ICMGGxxxx+000000 restart grib file has a record for the land-sea mask variable 172 while the regular records (e.g. ICMGGxxxx+201501) don't. This seems to lead to the grib filtering generating 2 timesteps for the land-sea mask in the first year:
`$~/wgrib/wgrib/wgrib 172.128.1.3
1:0:d=50010100:LSM:kpds5=172:kpds6=1:kpds7=0:TR=0:P1=0:P2=0:TimeU=1:sfc:anl:type=9:NAve=0
2:178296:d=50010100:LSM:kpds5=172:kpds6=1:kpds7=0:TR=0:P1=0:P2=0:TimeU=1:sfc:anl:type=9:NAve=0
`
While for the second year:
`$ ~/wgrib/wgrib/wgrib 172.128.1.3 
1:0:d=51010100:LSM:kpds5=172:kpds6=1:kpds7=0:TR=0:P1=0:P2=0:TimeU=1:sfc:anl:type=9:NAve=0
`

This seems to then cause different kinds of problems. Some files have two land-sea mask records right in the beginning which then causes the problem of unsupported file structure for cdo. Here for example for the variable tslsi:
`
$ ~/wgrib/wgrib/wgrib 235.128.1_172.128.1.3
1:0:d=15010100:LSM:kpds5=172:kpds6=1:kpds7=0:TR=0:P1=0:P2=0:TimeU=1:sfc:anl:type=9:NAve=0
2:178296:d=15010100:LSM:kpds5=172:kpds6=1:kpds7=0:TR=0:P1=0:P2=0:TimeU=1:sfc:anl:type=9:NAve=0
3:356592:d=15010100:SKT:kpds5=235:kpds6=1:kpds7=0:TR=0:P1=0:P2=0:TimeU=1:sfc:anl:type=9:NAve=0
4:534888:d=15010103:SKT:kpds5=235:kpds6=1:kpds7=0:TR=0:P1=0:P2=0:TimeU=1:sfc:anl:type=9:NAve=0
5:713184:d=15010106:SKT:kpds5=235:kpds6=1:kpds7=0:TR=0:P1=0:P2=0:TimeU=1:sfc:anl:type=9:NAve=0
...
`

For other variables (here snw) the result is a bit different which means that cdo can still read it:
`$ ~/wgrib/wgrib/wgrib 141.128.1_172.128.1.3
1:0:d=50010100:LSM:kpds5=172:kpds6=1:kpds7=0:TR=0:P1=0:P2=0:TimeU=1:sfc:anl:type=9:NAve=0
2:178296:d=50010100:SD:kpds5=141:kpds6=1:kpds7=0:TR=0:P1=0:P2=0:TimeU=1:sfc:anl:type=9:NAve=0
3:445430:d=50010100:LSM:kpds5=172:kpds6=1:kpds7=0:TR=0:P1=0:P2=0:TimeU=1:sfc:anl:type=9:NAve=0
4:623726:d=50010103:SD:kpds5=141:kpds6=1:kpds7=0:TR=0:P1=0:P2=0:TimeU=1:sfc:anl:type=9:NAve=0
5:890860:d=50010106:SD:kpds5=141:kpds6=1:kpds7=0:TR=0:P1=0:P2=0:TimeU=1:sfc:anl:type=9:NAve=0
...
`
While for the next year, there is only one record for LSM:
`$ ~/wgrib/wgrib/wgrib 141.128.1_172.128.1.3
1:0:d=51010100:LSM:kpds5=172:kpds6=1:kpds7=0:TR=0:P1=0:P2=0:TimeU=1:sfc:anl:type=9:NAve=0
2:178296:d=51010100:SD:kpds5=141:kpds6=1:kpds7=0:TR=0:P1=0:P2=0:TimeU=1:sfc:anl:type=9:NAve=0
3:445430:d=51010103:SD:kpds5=141:kpds6=1:kpds7=0:TR=0:P1=0:P2=0:TimeU=1:sfc:anl:type=9:NAve=0
4:712564:d=51010106:SD:kpds5=141:kpds6=1:kpds7=0:TR=0:P1=0:P2=0:TimeU=1:sfc:anl:type=9:NAve=0
5:979698:d=51010109:SD:kpds5=141:kpds6=1:kpds7=0:TR=0:P1=0:P2=0:TimeU=1:sfc:anl:type=9:NAve=0
...
`

For again other variables (here mrro), all the timesteps available for the variable get reduced to only two, causing a broken time series in the output:
`$ ~/wgrib/wgrib/wgrib 205.128.1.3_172.128.1.3 
1:0:d=50010100:RO:kpds5=205:kpds6=1:kpds7=0:TR=0:P1=0:P2=0:TimeU=1:sfc:anl:type=9:NAve=0
2:178296:d=50010100:LSM:kpds5=172:kpds6=1:kpds7=0:TR=0:P1=0:P2=0:TimeU=1:sfc:anl:type=9:NAve=0
3:356592:d=50010103:RO:kpds5=205:kpds6=1:kpds7=0:TR=0:P1=0:P2=0:TimeU=1:sfc:anl:type=9:NAve=0
4:534888:d=50010100:LSM:kpds5=172:kpds6=1:kpds7=0:TR=0:P1=0:P2=0:TimeU=1:sfc:anl:type=9:NAve=0
`

Further evidence is that there is no problem when linking the final record of the first year as previous month for the same.

**Comment by goord** [comment]
Hi @LaKenLin your analysis is correct. The change causing this has been an 'extra round' of grib filtering to get the static fields out of the initial state file. However, this has lead to a double counting (double writing) of those fields when you have linked your init state file to your previous month file like you have. I am fixing this by keeping track of grib records that were processed already

**Comment by LaKenLin** [comment]
Hi @goord, I just tried to explicitly remove code 172 from the initial state file and everything worked fine so your solution should do the trick for all my problems.
Thank you!

**Comment by goord** [comment]
I have pushed a [new branch](https://github.com/EC-Earth/ece2cmor3/tree/fx-filter-fix) that should fix your problem; @treerink should we merge this branch for the release?

**Comment by treerink** [comment]
@goord I probably do need to test this? In principle, yes it would be good to include it for the next v1.8 release. But #705 has priority, so because of that priority I would favour to merge the `sftlf-b2share` first, what do you think when do you have a next version to test for `sftlf-b2share`?

Ok I can run a test in parallel for the `fx-filter-fix` branch.

**Comment by LaKenLin** [comment]
In a first test the new branch runs without any problems.
Thank you, can be closed from my side!

**Comment by treerink** [comment]
Thanks @LaKenLin for reporting. I submitted my test-all case to check whether nothing is broken in that one.

**Comment by treerink** [comment]
The test-all tests where fine.

The `fx-filter-fix` branch is merged into the master.

So closing this issue.
